### ISSUE TITLE
testing.TestEnvironment not equal to itself

### ISSUE THREAD

**Issue by rickeylev** [bazelbuild/bazel#15225]
### Description of the bug:

The `testing.TestEnvironment` object is a bit strange: it isn't equal to itself. In comparison, most provider keys are equal to themselves.

This makes it difficult to write code that can properly detect when the testing.TestEnvironment provider key was requested, e.g.

```
def get_provider(key):
  if key == testing.TestEnvironment: ...
```
(in my particular case, I'm writing helper code for tests to assert TestEnvironment)

Also, I'm not sure if this should be TestEnvironment or TestEnvironmentInfo -- the public symbol is `testing.TestEnvironment`, but the docs say it returns `TestEnvironmentInfo` (note no leading "testing" namespace and the extra Info suffix).

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```
print(testing.TestEnvironment == testing.TestEnvironment)
```

actual: prints false

expected: prints true

### Which operating system are you running Bazel on?

linux

### What is the output of `bazel info release`?

google build

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

n/a

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

```text
n/a
```

### Have you found anything relevant by searching the web?

[15224 ](https://github.com/bazelbuild/bazel/issues/15224) is sort of related -- same category of "TestEnvironemt provider acts weird"

**Comment by brentleyjones** [comment]
Related, I would love a way to query for the `TestEnvironmentInfo` in a `Target`. This TODO seems important: https://github.com/bazelbuild/bazel/blob/ee9727df980f268bfb2b05637d199d55963fb6d8/src/main/java/com/google/devtools/build/lib/starlarkbuildapi/test/TestingModuleApi.java#L52

**Comment by fmeum** [comment]
I'm working on refactoring `TestEnvironmentInfo` into a proper exported `RunEnvironmentInfo` provider (name is TBD) that can be used for any executable target, not just tests. That should fix #15224, #15225 and #7364.

**Comment by brandjon** [comment]
It seems surprising to me that it wouldn't compare equal to itself, unless a new method descriptor is being created each time the method is retrieved. But I thought we cached those.
